src/Provider/DecoratingCacheSettingsProvider.php

```diff
@@ -0,0 +1,342 @@
+
+        if ($cacheItem->isHit()) {
+            return $this->deserializeArray(array_filter($cacheItem->get()));
+        }
+
+        $domains = $this->decoratingProvider->getDomains($onlyEnabled);
+
+        if (!empty($domains)) {
+            $this->storeCached($cacheItem, $this->serializeArray($domains));
+        }
+
+        return $domains;
+    }
+
+    private function serializeArray(array $elements): array
+    {
+        foreach ($elements as &$element) {
+            $element = $this->serializer->serialize($element, 'json');
+        }
+
+        return $elements;
+    }
+
+    private function deserializeArray(array $elements): array
+    {
+        foreach ($elements as &$element) {
+            $element = $this->serializer->deserialize($element, DomainModel::class, 'json');
+        }
+
+        return $elements;
+    }
+
+    public function isReadOnly(): bool
+    {
+        return $this->decoratingProvider->isReadOnly();
+    }
+
+    public function save(SettingModel $settingModel): bool
+    {
+        return $this->decoratingProvider->save($settingModel);
+    }
+
+    public function delete(SettingModel $settingModel): bool
+    {
+        return $this->decoratingProvider->delete($settingModel);
+    }
+
+    public function updateDomain(DomainModel $domainModel): bool
+    {
+        return $this->decoratingProvider->updateDomain($domainModel);
+    }
+
+    public function deleteDomain(string $domainName): bool
+    {
+        return $this->decoratingProvider->deleteDomain($domainName);
+    }
+
+    private function getCached(string $key): CacheItem
+    {
+        return $this->cache->getItem($key);
+    }
+
+    private function storeCached(CacheItem $cacheItem, $value, bool $commit = true): void
+    {
+        $cacheItem->set($value);
+        $this->cache->saveDeferred($cacheItem);
+
+        if (true === $commit) {
+            $this->cache->commit();
+        }
+    }
+
+    private function getSettingNamesKey(string $domainName): string
+    {
+        return sprintf('setting_names[%s]', $domainName);
+    }
+
+    private function getSettingKey(string $domainName, string $settingName): string
+    {
+        return sprintf('setting[%s][%s]', $domainName, $settingName);
+    }
+
+    private function getDomainKey(bool $onlyEnabled = false): string
+    {
+        return sprintf('domain%s', $onlyEnabled ? '_oe' : '');
+    }
+
+    private function clearIfNeeded(): void
+    {
+        $lastCheck = $this->cache->getItem('last_modification_time_check');
+        $time = time();
+        if ($lastCheck->isHit() && time() - $lastCheck->get() < $this->checkValidityInterval) {
+            return;
+        }
+
+        if ($this->getModificationTime() < $this->decoratingProvider->getModificationTime()) {
+            $lock = $this->lockFactory->createLock(__FUNCTION__);
+
+            if (!$lock->acquire()) {
+                usleep(self::LOCK_RETRY_INTERVAL_MS);
+                $this->clearIfNeeded();
```

Comment:
shouldn't you `return;` after this? now with each unsuccessful acquire it will go deeper and when eventually succeed cache will be cleared how many times it was unsuccessful.

Reply:
:facepalm: fixed now for all locks